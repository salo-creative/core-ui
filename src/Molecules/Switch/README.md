# Switch

A simple switch/toggle component

---

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
import Switch from '@salo/core-ui/Molecules/Switch';
```

Implement as follows

```javascript
<Switch
  disabled={ false } // Set if you want the input disabled
  label='On' // Set if you want a label
  labelOff='Off' // if you want different label for off state
  loading={ false } // Set if you want to trigger the loading state
  onChange={ (value) => setState(value) }
  size={ size }
  value={ state }
/>
```

**For full prop types and usage see storybook info/knobs**

