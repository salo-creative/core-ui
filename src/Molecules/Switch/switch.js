import React from 'react';
import PropTypes from 'prop-types';

// COMPONENTS & STYLES
import Loader from '../Loader';
import {
  SwitchWrapper,
  SwitchContainer,
  SwitchHandle,
  SwitchLabel,
  LoaderContainer
} from './switch.styles';

const Switch = (props) => {
  const {
    className,
    disabled,
    justify,
    label,
    labelOff,
    labelPosition,
    loading,
    margin,
    name,
    onChange,
    showBoth,
    size,
    transition,
    value,
    width
  } = props;

  const isDisabled = disabled || loading;

  const renderLabel = () => {
    if (showBoth) return label;
    if (!value && labelOff) {
      return labelOff;
    }
    return label;
  };

  return (
    <SwitchWrapper
      justify={ justify }
      disabled={ isDisabled }
      size={ size }
      margin={ margin }
      className={ `${ className } salo-switch ${ labelPosition === 'right' ? 'alignRight' : 'alignLeft' }` }
      width={ width }
    >
      { showBoth && (
        <SwitchLabel
          className={ `salo-switch__label left ${ !value ? 'active' : '' }` }
          size={ size }
          htmlFor={ name }
        >
          { labelOff || label }
        </SwitchLabel>
      ) }
      <SwitchContainer
        active={ value }
        onClick={ () => onChange(!value) }
        size={ size }
        id={ name }
        role='switch'
        ariaChecked={ `${ !!value }` }
        className={ `salo-switch__container salo-switch__container--${ value ? 'active' : 'inactive' }` }
      >
        <SwitchHandle
          active={ value }
          size={ size }
          className='salo-switch__handle'
          hasTransition={ transition }
        />
        { loading && (
          <LoaderContainer
            size={ size }
            active={ value }
            className='salo-switch__loader'
          >
            <Loader
              display
              appearance='light'
              loaderProps={ {
                size: 25
              } }
            />
          </LoaderContainer>
        ) }
      </SwitchContainer>
      <SwitchLabel
        className={ `salo-switch__label ${ value ? 'active' : '' }` }
        size={ size }
        htmlFor={ name }
      >
        { renderLabel() }
      </SwitchLabel>
    </SwitchWrapper>
  );
};

Switch.defaultProps = {
  className: '',
  disabled: false,
  justify: '',
  label: '',
  labelOff: '',
  labelPosition: 'right',
  loading: false,
  margin: '0',
  showBoth: true,
  size: 'M',
  transition: 'left 0.3s ease-in-out',
  value: false,
  width: 'auto'
};

Switch.propTypes = {
  className: PropTypes.string,
  disabled: PropTypes.bool,
  justify: PropTypes.string,
  label: PropTypes.string,
  labelOff: PropTypes.string,
  labelPosition: PropTypes.oneOf(['left', 'right']),
  loading: PropTypes.bool,
  margin: PropTypes.string,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  showBoth: PropTypes.bool,
  size: PropTypes.oneOf(['L', 'M']),
  transition: PropTypes.string,
  value: PropTypes.bool,
  width: PropTypes.string
};

export default Switch;