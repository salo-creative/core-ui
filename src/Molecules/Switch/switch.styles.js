import styled from 'styled-components';
import { transparentize } from 'polished';

export const SwitchWrapper = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  height: ${ ({ size }) => (size === 'L' ? '4rem' : '4rem') };
  margin: ${ ({ margin }) => margin };
  position: relative;
  opacity: 1;
  ${ ({ width }) => (width ? `width: ${ width };` : '') }
  ${ ({ justify }) => (justify ? `justify-content: ${ justify };` : '') }
  ${ ({ disabled }) => (disabled ? `
    cursor: not-allowed;
    opacity: 0.75;
    &:after {
      content: '';
      position: absolute;
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
    }
  ` : 'cursor: pointer;') }
  &.alignLeft {
    flex-direction: row-reverse;
  }
`;

export const SwitchContainer = styled.button`
  display: flex;
  width: 5rem;
  height: ${ ({ size }) => (size === 'L' ? '2rem' : '1.5rem') };
  position: relative;
  border-radius: 0.4rem;
  background: ${ ({ active, theme }) => (active ? transparentize(0.7, theme.success) : theme.grey) };
  cursor: pointer;
  margin: 0 1rem 0 0;
  border: none;
  outline: none;
  .alignLeft & {
    margin: 0 0 0 1rem;
  }
`;

export const SwitchHandle = styled.span`
  top: -0.5rem;
  position: absolute;
  border-radius: 50%;
  cursor: pointer;
  border: 0.2rem solid #fff;
  background: ${ ({ active, theme }) => (active ? theme.success : theme.darkGrey) };
  transition: ${ ({ hasTransition }) => hasTransition };
  ${ ({ size, active }) => (size === 'L' ? `
    height: 3rem;
    width: 3rem;
    left: ${ (active ? '2rem' : 0) };
  ` : `
    height: 2.5rem;
    width: 2.5rem;
    left: ${ (active ? '2.5rem' : 0) };
  `) }
`;

export const SwitchLabel = styled.span`
  color: ${ ({ theme }) => theme.darkGrey };
  font-size: ${ ({ size }) => (size === 'L' ? '1.6rem' : '1.4rem') };
`;

export const LoaderContainer = styled.div`
  position: absolute;
  top: -0.5rem;
  ${ ({ size, active }) => (size === 'L' ? `
    top: -0.25rem;
    left: ${ (active ? '2.25rem' : '0.25rem') };
  ` : `
    top: -0.5rem;
    left: ${ (active ? '2.5rem' : 0) };
  `) }
`;