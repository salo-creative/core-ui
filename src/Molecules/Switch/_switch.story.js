import React from 'react';
import { storiesOf } from '@storybook/react';
import {
  withKnobs, select, text, boolean
} from '@storybook/addon-knobs';
import { State, Store } from '@sambego/storybook-state';
import styled from 'styled-components';

// load tests
import { withTests } from '@storybook/addon-jest';
import results from '../../../.storybook/jest-test-results.json';

// FEATURED COMPONENT
import { Switch } from '../../index';

// README //
import README from './README.md';

// Start of story logic
const stories = storiesOf('Molecules | Switch', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withTests({
  results
}));
stories.addParameters({
  jest: ['switch']
});

const CustomSwitch = styled(Switch)`
  border: 1px dashed hotpink;
  border-radius: 25px;
  padding: 0 10px;

  .salo-switch__container {
    transition: background-color ${ ({ transitionTime }) => transitionTime } ease-in;
  }

  .salo-switch__container--active {
    background: rebeccapurple;
  }

  .salo-switch__handle {
    background: slateblue;
  }

  .salo-switch__label {
    color: #444;
  }
`;

stories.add(
  'Basic',
  (() => {
    const disabled = boolean('disabled', false);
    const label = text('label', 'Yes');
    const labelOff = text('labelOff', 'No');
    const labelPosition = select('label position', ['left', 'right'], 'right');
    const loading = boolean('loading', false);
    const showBoth = boolean('showBoth', false);
    const size = select('size', ['L', 'M'], 'M');
    // STORE
    const store = new Store({
      value: false
    });
    return (
      <State store={ store }>
        { state => (
          <Switch
            disabled={ disabled }
            label={ label }
            labelOff={ labelOff }
            labelPosition={ labelPosition }
            loading={ loading }
            showBoth={ showBoth }
            onChange={ (value) => store.set({
              value
            }) }
            size={ size }
            value={ state.value }
          />
        ) }
      </State>
    );
  }),
  {
    info: {
      propTables: [Switch],
      propTablesExclude: [State]
    },
    notes: README
  }
);

stories.add(
  'Customised',
  (() => {
    const disabled = boolean('disabled', false);
    const label = text('label', 'Stuff');
    const labelOff = text('labelOff', '');
    const labelPosition = select('label position', ['left', 'right'], 'right');
    const loading = boolean('loading', false);
    const size = select('size', ['L', 'M'], 'M');
    // STORE
    const store = new Store({
      value: false
    });
    return (
      <State store={ store }>
        { state => (
          <CustomSwitch
            disabled={ disabled }
            label={ label }
            labelOff={ labelOff }
            labelPosition={ labelPosition }
            loading={ loading }
            onChange={ (value) => store.set({
              value
            }) }
            size={ size }
            transitionTime='1s'
            value={ state.value }
          />
        ) }
      </State>
    );
  }),
  {
    info: {
      propTables: [Switch],
      propTablesExclude: [State]
    },
    notes: README
  }
);