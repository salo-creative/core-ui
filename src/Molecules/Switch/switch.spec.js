import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import Switch from './index';

test('Renders against snapshot', async () => {
  const { container } = renderWithTheme(
    <Switch />
  );
  expect(container.firstChild).toMatchSnapshot();
});