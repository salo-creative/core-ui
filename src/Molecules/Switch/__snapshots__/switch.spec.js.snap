// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 4rem;
  margin: 0;
  position: relative;
  opacity: 1;
  width: auto;
  cursor: pointer;
}

.c0.alignLeft {
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 5rem;
  height: 1.5rem;
  position: relative;
  border-radius: 0.4rem;
  background: #e9ecef;
  cursor: pointer;
  margin: 0 1rem 0 0;
  border: none;
  outline: none;
}

.alignLeft .c1 {
  margin: 0 0 0 1rem;
}

.c2 {
  top: -0.5rem;
  position: absolute;
  border-radius: 50%;
  cursor: pointer;
  border: 0.2rem solid #fff;
  background: #6c757d;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
  height: 2.5rem;
  width: 2.5rem;
  left: 0;
}

.c3 {
  color: #6c757d;
  font-size: 1.4rem;
}

<div
  class="c0 alignRight"
  width="auto"
>
  <button
    class="c1"
  >
    <span
      class="c2"
    />
  </button>
  <span
    class="c3"
  />
</div>
`;
